packages/block-library/src/legacy-widget/edit/preview.js

```diff
@@ -6,13 +6,44 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
+import { useRefEffect } from '@wordpress/compose';
 import { addQueryArgs } from '@wordpress/url';
 import { useState } from '@wordpress/element';
 import { Placeholder, Spinner, Disabled } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 
 export default function Preview( { idBase, instance, isVisible } ) {
-	const [ iframeHeight, setIframeHeight ] = useState( null );
+	const [ iframeHeight, setIframeHeight ] = useState();
+
+	// Resize the iframe on either the load event, or when the iframe becomes visible.
+	const ref = useRefEffect( ( iframe ) => {
+		function onChange() {
+			const boundingRect = iframe?.contentDocument?.body?.getBoundingClientRect();
+			if ( boundingRect ) {
+				// Include `top` in the height calculation to avoid the bottom
+				// of widget previews being cut-off. Most widgets have a
+				// heading at the top that has top margin, and the `height`
+				// alone doesn't take that margin into account.
+				setIframeHeight( boundingRect.top + boundingRect.height );
+			}
+		}
+
+		const { IntersectionObserver } = iframe.ownerDocument.defaultView;
+
+		// Observe for intersections that might cause a change in the height of
+		// the iframe, e.g. a Widget Area becoming expanded.
+		const intersectionObserver = new IntersectionObserver( onChange, {
```

Comment:
Should we also call `intersectionObserver.disconnect()` to free up the memory?

Reply:
Sounds good. First time using `intersectionObserver`, so wasn't aware it needed to be `disconnect`ed.

Reply:
What does `IntersectionObserver` actually do? There seem to be no docs for it and the name confuses the heck out of me. I don't understand what an intersection might be in this context 😅

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

It's usually used for lazily showing stuff that's scrolled into view, but also works in this case. The idea is to observe for when the block preview element becomes visible (the widget area is expanded) and only calculate the height of the content when that happens. This was the first time I've had the chance to use it because I've always had to support IE before. 😄

Reply:
Oooh nice, I didn't realise it was an actual web API!